cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -580,12 +582,30 @@ public SearchResults apply(MetadataDataset input) throws Exception {
   }
 
   @Override
-  public void rebuildIndexes(MetadataScope scope) throws DatasetManagementException, IOException {
+  public void rebuildIndexes(MetadataScope scope,
+                             RetryStrategy retryStrategy) {
     byte[] row = null;
-    while ((row = rebuildIndex(row, scope)) != null) {
+    while ((row = rebuildIndexesWithRetries(scope, row, retryStrategy)) != null) {
       LOG.debug("Completed a batch for rebuilding {} metadata indexes.", scope);
     }
+  }
 
+  private byte[] rebuildIndexesWithRetries(final MetadataScope scope,
```

Comment:
Not sure if I understand this one. rebuildIndexes() is not throwing DatasetManagementException and IOException. 
rebuildIndexesWithRetries() has a callWithRetries which will underneath retry all Retryable Exceptions. Anything else is wrapped into Runtime Exception.